### ISSUE TITLE
[Feature Request] Implement `copy` function

### ISSUE THREAD

**Issue by nemo83** [Hyperion-BT/helios#60]
Given the functional nature of helios, it would be amazing a if `copy` function could be implemented for onchain types. This would make helios less error prone by reducing verbosity.

Example. 

We have a combination of lovelace/tokens at a utxo. We know that the value it's legit but we don't care what's in it, neither we want to keep track, and we want to split into two parts:
1. All the tokens + fixed amount of ada to be sent to an address/script
2. Whatever is left (fee/costs) to be sent to a wallet.

```  javascript
func correct_value_sent(
    tx: Tx, 
    totalValueLocked: Value,
    vaultPkh: PubKeyHash, 
    walletPkh: PubKeyHash
    ) -> Bool {

    valueSentToVault: Value = totalValueLocked.copy(lovelaces = 2000000)
    costs: Value = totalValueLocked - vaultValue
    
    walletValue: Value = tx.value_sent_to(walletPkh);
    vaultValue: Value = tx.value_sent_to_datum(vaultPkh, MyDatumDefinition, true);

    (costs >= walletValue) && (valueSentToVault == vaultValue)
  }
```

**Comment by christianschmitz** [comment]
Not sure if setters should be auto-generated for user types (eg. `set_<name-of-field>`)

**Comment by christianschmitz** [comment]
We can add `copy` as special syntax (similar to how `switch` is special syntax)

**Comment by nemo83** [comment]
> Anything that you can think of that can be improved upon wrt. scala?
Not sure what you mean here, but the `copy` method is the only scala syntax feature I can think of that could be beneficial to helios. The other thing I use all the time is pattern matching (switch), that helios has already.

Wrt to alternative to the `copy` I would avoid the `set` any other thing sounds good to me.

**Comment by christianschmitz** [comment]
Variations on the syntax are possible:

```
my_struct.copy(field1 = <expr1>, field2 = <expr1>)
```

or 

```
my_struct.copy{field1 = <expr1>, field2 = <expr1>}
```

or

```
my_struct.copy{field1 => <expr1>, field2 => <expr1>}
```

or

```
my_struct.copy{field1: <expr1>, field2: <expr1>}
```

**Comment by christianschmitz** [comment]
I think the last one is most consistent with the rest of Helios. @nemo83 What do you think?

**Comment by nemo83** [comment]
> I think the last one is most consistent with the rest of Helios. @nemo83 What do you think?

I am _used_ to the first example. But I'd be interested in seeing examples of where 4 is already used in helios.

Thanks 🙏

**Comment by christianschmitz** [comment]
An issue I see with using scala `copy` syntax is needing to expand key-based arguments to the rest of the language, otherwise, it feels incomplete.

My reasoning for the braces syntax is that it is like a literal struct, except it does not need all the fields.

**Comment by christianschmitz** [comment]
Maybe expanding key-based arguments to the rest of the language is not such a bad idea, and also add the capability for optional arguments.

Perhaps we could use the C# syntax ?:

```
myValue.copy(lovelace: 200000)
```

**Comment by christianschmitz** [comment]
Actually, now that I think of it: we have to use the colon syntax because the equals syntax is already a part of a valid expression (and function call arguments can be arbitrary expressions)

**Comment by christianschmitz** [comment]
What should the type signature look like for functions with optional arguments?

(I don't think I've ever seen a language do that right)

**Comment by nemo83** [comment]
> What should the type signature look like for functions with optional arguments?
> 
> Eg.
> 
> ```
> (a: Int, b? Int) -> Int
> ```
> 
> Which could be cast into `(a: Int) -> Int` or `(a: Int, b: Int) -> Int` automatically, or even `(Int) -> Int` or `(Int, Int) -> Int` (arg names are now part of the type)
> 
> (I don't think I've ever seen a language do that right)

This is a very delicate topic, and in this case I can see now why in a way `copy` is different.

For me optionality should still be wrapped into the `Option[]` type.

So the `:?` syntax would/should be only for `copy` method as it won't make sense elsewhere.

**Comment by christianschmitz** [comment]
To be clear: `copy` would be a special builtin function that wouldn't be usable as a value (it would have to be called immediately).

But I think I figured out the optimal type syntax for functions with optional args: `(a: Int; b: Int, c: Int) -> Int` (so anything after `;` is optional)

**Comment by christianschmitz** [comment]
Another variation in which the type signature differs from the definition:
```
func(a: Int, b: Int = 0) -> Int {
   ...
}

// func(a: Int, b: ?Int) -> Int
// func(Int, ?Int) -> Int
```

`copy` could actually be an auto-generated function:
```
myStruct.copy(<name1>: ?Type1, <name2>: ?Type2, ...) -> MyStruct
```

**Comment by christianschmitz** [comment]
I've implemented default values and `copy` in v0.13.0 of Helios. @nemo83  When you have some time, could you test that it works for your cases?

**Comment by nemo83** [comment]
Huge milestone for Helios.

Less code
Less bugs

#LFG
